crates/runtime/src/embeddings/execution_plan.rs

```diff
@@ -314,24 +314,46 @@ pub(crate) async fn compute_additional_embedding_columns(
 /// +---------------------+      +-------------+
 ///     [`StringArray`]        [`FixedSizeListArray`]
 /// ```
+#[allow(clippy::cast_possible_truncation, clippy::cast_possible_wrap)]
 async fn get_vectors(
     arr: impl Iterator<Item = Option<&str>>,
     model: &dyn Embed,
 ) -> Result<FixedSizeListArray, Box<dyn std::error::Error + Send + Sync>> {
-    let field = Arc::new(Field::new("item", DataType::Float32, false));
-    let column: Vec<String> = arr.filter_map(|s| s.map(ToString::to_string)).collect();
+    // Filter out nulls or empty strings before caling [`Embed::embed`].
+    let (null_pairs, values): (Vec<_>, Vec<_>) = arr
+        .enumerate()
+        .partition(|(_, o)| o.is_none() || o.is_some_and(|v| !v.is_empty()));
+    let nulls: Vec<usize> = null_pairs.into_iter().map(|(i, _)| i).collect();
+
+    let column: Vec<String> = values
+        .iter()
+        .filter_map(|(_, s)| s.map(ToString::to_string))
+        .collect();
+
     let embedded_data = model.embed(EmbeddingInput::StringArray(column)).await?;
     let vector_length = embedded_data.first().map(Vec::len).unwrap_or_default();
-    let processed = embedded_data.iter().flatten().copied().collect_vec();
 
-    let values = Float32Array::try_new(processed.into(), None)?;
-    FixedSizeListArray::try_new(
-        Arc::clone(&field),
-        i32::try_from(vector_length)?,
-        Arc::new(values),
-        None,
-    )
-    .boxed()
+    // Add back nulls into the ordering of embedding data.
+    let mut builder = FixedSizeListBuilder::with_capacity(
+        PrimitiveBuilder::<Float32Type>::with_capacity(
+            (embedded_data.len() + nulls.len()) * vector_length,
+        ),
+        vector_length as i32,
+        embedded_data.len() + nulls.len(),
+    );
+
+    let mut i: usize = 0;
+    for vector in embedded_data {
```

Comment:
Interesting approach could be constructing resultant array by providing offsets and logical nulls
https://github.com/spiceai/spiceai/blob/67879865616df4197290d9ad2d7a9be5c09d3300/crates/arrow_tools/src/record_batch.rs#L124

Reply:
They key reason we can't do this is because we split out empty/null inputs, from the remaining and only pass non-empty/null inputs to the model (`let embedded_data = model.embed(EmbeddingInput::StringArray(column)).await?;`). We then need to recombine it, we don't have a contiguous block of data.

Reply:
Confirm this is to recombine: the idea is to just specify nulls/offests instead of `append`/`append_nulls`

Reply:
I also think we should cover this by unit test to ensure this is not broken at some point